Merge the class and rt-class attributes

anchann opened the pull request:
Hello! I have a number of PRs, and wasn't sure where to put the high level description, so I'll copy and paste it in all of them.

Coming from Angular, I found the react world to be a sad place as far as DSLs go — JSX is a sorry excuse that for some reason beyond me is accepted by the community as the standard. And then I stumbled upon your react-templates work. Thank you!! While converting my work over, I came across a number of wishlist features, and today finally decided to take some time to implement them. They are all aimed at improving readability/writability, as well as sometimes avoiding unnecessary nested elements.

===


Previously an error was raised when a node had both the class and the
rt-class attribute defined. This leads to unnecessary verbose code which
taxes the reader. Splitting static class names from dynamic ones removes
this distraction, and keeps the semantics clean in the reader's mind.

On renaming Props to Attrs. I would have liked to avoid such global
renames on my first commit to somebody else's code base, but I needed to
add a constant for "className", and that seemed like a property on a
React object, as opposed to an attribute on an html element that only
later gets converted to a property. So I took the liberty to rename the
variables.

On tests. Having noticed how order-depenedent some of the code is (in
particular when it comes to dealing with rt-if/rt-repeat/rt-scope), I
decided it might be a good idea to be somewhat black-boxy in the tests,
and ensure that merging of class names works correctly in both
directions.

brodybits commented:
@anchann it looks like you have some nice improvements. I do have the following comments:

It looks like this change may break some existing templates or applications, which we want to avoid at almost all cost. For this change, I recommend that `rt-class` should simply be deprecated (until a certain point).

One thing this project is missing is an ability to define new attributes, which should be defined as some form of plugins.

When I look at the code history, issues, and PRs the owners do not seem to be responsive anymore. Perhaps someone can start a new fork to continue the work.

You may want to look at #21, where someone wanted repeat-inside functionality. I started working on a solution for it, as you can see in the discussion. I think this should be in a plugin.

I may come back to this, if a have a chance, someday.

anchann commented:
Hey @brodybits,

> deprecating `rt-class`

Did you mean to say rt-scope? There's nothing breaking in my changes to `rt-class`: previously, having both `class` and `rt-class` resulted in an error, and didn't work at all, and now it works. `class` by itself, as well as `rt-class` by itself, work as before.

> avoiding breaking changes

Agreed. While the code is fresh in my head, I'll break out the new functionality into something like `rt-inner-scope`, and leave `rt-scope` alone, marking it with a warning in the docs.

> plugins

Hmm.

From the practical standpoint, I definitely don't have the time to re-architect the code to make plugins a reality, and as-is there's no way the code can support them, given how dependent on order of execution all the current directives are.

From the theoretical standpoint, I'm not convinced that plugins are the right way to go. Unlike angular, where attributes represent both control structures and library features, react-template attributes are strictly the former. In that sense, while it would be natural to have a way to plug in new library modules, it feels far less natural to be able to plug in new control structures at will. People with a Lisp background would disagree, since they build new control structures all the time, but Lisp works on general lists, while here we have a highly restricted tree data structure. So, I feel that the handful of additional useful control structure additions are most likely to come from just people forking the repo, understanding the code deeply, and then implementing the new structures. That is to say I think that the number of general high quality control structures contributed with or without the plugin framework would be about the same, which would not justify the effort in developing the framework.

> owners not responsive

That's unfortunate. I personally feel that my improvements make react-templates more of a pleasure to use (I built the improvements for myself, really; sharing was secondary). After making the unbreaking/deprecating changes for rt-scope that you suggested, I'd be happy to publish the project as a new npm module, perhaps something like react-templates-improved, and mark it as deprecated if owners of the main branch start responding again. I wouldn't mind being the maintainer for a while — the mental calisthenics that doing metaprogramming forces me through seem healthy every once in a while.

Thoughts?

brodybits commented:
> Thoughts?

I just looked at the history and the project does seem to remain active, though they are a bit slow and also slow to respond. We _could_ keep the changes in a `react-templates-extended` or something while we wait for them to look at the changes (perhaps another prefix for tags that are not yet accepted). Also, if a project is really not maintained it _is_ possible to take the npm module over: https://docs.npmjs.com/misc/disputes

anchann commented:
I made a new branch, put all my changes there, updated documentation, and implemented rt-repeat-children.

https://github.com/anchann/react-templates

danyshaanan commented:
Hi @anchann && @brodybits. Please see my comment here: https://github.com/wix/react-templates/issues/49#issuecomment-125505257
